Design document for Test Skip Expiry Feature

opcoder0 opened the pull request:
### Description of PR

The PR defines the design document for implementing a new feature on the conditional mark plugin. The feature allows users to specify an expiry date on the skipped / xfail-ed test cases.

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] New Test case
    - [ ] Skipped for non-supported platforms
- [x] Test case improvement


### Back port request
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405
- [ ] 202411
- [ ] 202505

### Approach
#### What is the motivation for this PR?

New feature. See description above.

#### How did you do it?

NA

#### How did you verify/test it?

NA

#### Any platform specific information?

NA

#### Supported testbed topology if it's a new test case?

NA

### Documentation

NA

opcoder0 committed:
Design document for Test Skip Expiry Feature

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

mssonicbld commented:
/azp run

StormLiangMS commented on docs/test_skip_expiry_design.md:
````diff
@@ -0,0 +1,278 @@
+```yaml
+acl/test_acl.py:
+  skip:
+    reason: "Skip acl for isolated-v6 topology"
````
Should we have a category why it get skipped? We may allow infinite skip time for specified skip reasons, for example, ASIC not support or TOPO not support? 
For any other skip, it should be skipped by next six months or less, and we may enforce this in PR checker? If someone want to skip, and not fall into the infinite skip categories, if the expiry_date is missing or longer than allowed time, it will be marked as PR checker failure. 
And how we handle if there are two platforms, for same case, one of them can apply infinite skip because the feature is not supported, but the other need to skip with expiry_date, could the skip plugin treat them as two different skip cases and honor both of them but not only one of them?

opcoder0 commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,278 @@
+
+**Recommendations**:
+1. **Warning Period**: Add `expiry_action: warn` with 1-2 week grace period
+2. **Automated Alerts**: Trigger alerts when skips are expiring soon
```
Thanks @ZhaohuiS. I was thinking of creating a GitHub workflow that runs only for master/main branch every week or so and checks for all tests that are expiring in the next 30, 14, 7, 1 or 0 days. About alerting the easier method could be to create a GitHub Issue and assign it to an owner to examine. I think alerting on Teams is possible, but I am not sure how to get that to work yet.

ZhaohuiS commented:
[like]  Zhaohui Sun reacted to your message:
________________________________
From: Sai Kiran ***@***.***>
Sent: Wednesday, December 3, 2025 3:01:29 AM
To: sonic-net/sonic-mgmt ***@***.***>
Cc: Zhaohui Sun ***@***.***>; Mention ***@***.***>
Subject: Re: [sonic-net/sonic-mgmt] Design document for Test Skip Expiry Feature (PR #21495)


@opcoder0 commented on this pull request.

________________________________

In docs/test_skip_expiry_design.md<https://github.com/sonic-net/sonic-mgmt/pull/21495#discussion_r2583464890>:

> +    reason: "Skip on platform A"
+    expiry_date: "2025-03-01"
+    conditions: ["platform == 'A'"]
+  xfail:
+    reason: "Known to fail on platform B"
+    expiry_date: "2025-06-01"
+    conditions: ["platform == 'B'"]
+```
+
+**Behavior**: Each mark evaluated independently with its own expiry
+
+### 6. Expired Skip in CI/CD
+
+**Recommendations**:
+1. **Warning Period**: Add `expiry_action: warn` with 1-2 week grace period
+2. **Automated Alerts**: Trigger alerts when skips are expiring soon


Thanks @ZhaohuiS<https://github.com/ZhaohuiS>. I was thinking of creating a GitHub workflow that runs only for master/main branch every week or so and checks for all tests that are expiring in the next 30, 14, 7, 1 or 0 days. About alerting the easier method could be to create a GitHub Issue and assign it to an owner to examine. I think alerting on Teams is possible, but I am not sure how to get that to work yet.

—
Reply to this email directly, view it on GitHub<https://github.com/sonic-net/sonic-mgmt/pull/21495#discussion_r2583464890>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AWRZHDVJTMMH46IEMIXE3J337ZHATAVCNFSM6AAAAACNUYOQBWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTKMZSHE4DIMRQGU>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

opcoder0 committed:
Add category based skip mechanism to existing design

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

StormLiangMS commented on docs/test_skip_expiry_design.md:
````diff
@@ -0,0 +1,465 @@
+#### Structure Example
+
+```yaml
+# Define skip categories at the top of the file
````
Should we maintain the categories in a separate file and restrict editing to specific roles? This would allow tighter control over category changes.

StormLiangMS reviewed (dismissed):
LGTM, with 2 minor questions.

opcoder0 commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,465 @@
+
+---
+
+### **Recommendation: Approach 1 (Inline Category and Expiry Date Fields)**
```
Yes, enforcement would happen in the PR checker during Pytest collection phase. To begin with the entries are backward compatible to allow gradual introduction of categories and dates. Once all dates / categories are set in the file we can turn on enforcement which wouldn't allow future additions / modifications without date for temporary skips.

opcoder0 committed:
Add design and specification for GH issue based temporary skip

* Adds design and description (as Approach 3) for using GitHub
  issue to specify temporary skips.

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

wangxin commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,1083 @@
+
+### Approach 3: GitHub Issue-Based Expiry Management
+
+Leverage GitHub issues as the source of truth for temporary skip expiry, with automated workflows managing issue lifecycle and expiry detection.
```
IMO, Github issue based expiry is correct direction. For a test that needs to be skipped, it could because of either permanent or temporary reason. For **temporary** reason, it means that something will be changed or fixed in the future. Then use a Github issue to track the future change makes the most sense.

The original motivation of this design is that some skips are always there and never been taken care of. These skips are only related with temporary skip. So, the real problem is that some Github issues should not be forgotten. We need some workflow/process to get necessary attention from people to get the issue fixed.

wangxin commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,1083 @@
+
+#### Disadvantages
+
+- **Split context**: Skip definitions in YAML, skip reasons/status in GitHub - two places to check
```
Github issue is usually not the only condition for skipping. So the skip reason in YAML is for all the conditions of a skip entry, not just for the Github issue.

wangxin commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,1083 @@
+      - "INFRASTRUCTURE_ISSUE"
+
+# Expiry configuration
+expiry_config:
```
Suggest to use a separate file for expiry config. The expiry config is used by the pipeline for expiring issues, not used by conditional mark plugin.

wangxin commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,1083 @@
+    allowed_reasons:
+      - "BUG_FIX_IN_PROGRESS"
+      - "NEW_FEATURE_UNDER_DEVELOPMENT"
+      - "INFRASTRUCTURE_ISSUE"
```
As I mentioned above, it is not necessary to have these categories.

wangxin commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,1083 @@
+
+To handle skips across multiple release branches:
+
+**Recommended Approach: Branch-Specific Labels**
```
I prefer per branch issue method. Using labels could have extra maintenance effort.

wangxin commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,1083 @@
+    reason: "BGP issue on main branch"
+    category: "BUG_FIX_IN_PROGRESS"
+    issue: "https://github.com/sonic-net/sonic-mgmt/issues/12345"
+    branches: ["main", "202311"]  # Issue applies to these branches
```
Do not add extra field `branches`. The branches check could be embedded in the conditions.

opcoder0 replied:
Updated. Yes. That won't be applicable anymore as the chosen method is going to be cloning issues for each release branch.

opcoder0 committed:
Address comments

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

opcoder0 commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,1083 @@
+acl/test_acl.py:
+  skip:
+    reason: "Skip acl for isolated-v6 topology"
+    category: "TOPO_NOT_SUPPORTED"  # Permanent - no issue needed
```
Thanks @wangxin. I agree we can do away with the temporary and permanent categories and use the implicit assumption that if a skip case has an issue associated with it then it is temporary otherwise it is permanent. However, I think having these labels may not be tied to a category. These labels could be used for reporting which will give us a quick overview of how many issues fall under a specific category type (like FLAKY_ISSUES, BUG_FIX_IN_PROGRESS or TOPO_NOT_SUPPORTED etc.).

opcoder0 committed:
Updated design document with latest agreements from review comments

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

mssonicbld commented:
/azp run

opcoder0 committed:
Update document - cleanup

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

ZhaohuiS commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,517 @@
+├─────────────────────────────────────────────────────────────────────────────┤
+│                                                                             │
+│  ┌─────────────────┐    ┌──────────────────┐    ┌─────────────────────┐   │
+│  │ tests_mark_     │───▶│ Conditional Mark │───▶│ GitHub Issue Cache  │   │
```
The "Chosen Expiry Strategy" section (line 253) picks Strategy B (priority escalation), but here the Implementation Approach says "Strategy A (time-based expiry) will be used initially." Which one is actually being built? If both — what's the transition trigger from A to B?

opcoder0 replied:
Thank you for spotting that. I have updated the document. This design uses priority tag-based escalation method to identify and track tests being skipped.

ZhaohuiS commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,517 @@
+#### YAML Structure
+
+There is no structural change with this design or approach.
+
```
Permanent = "no GitHub issue", temporary = "has GitHub issue" — but plenty of permanent skips today have tracking issues (e.g., ASIC_NOT_SUPPORTED with an issue explaining the HW limitation). Wouldn't this silently reclassify them as temporary and start the expiry clock?

opcoder0 replied:
Yes, that's correct. If we spot such issues, we need to clean up the conditional mark file and keep only those issues that are temporary in nature and document cases which are permanently skipped or move permanent skips to a separate conditional mark file just for clarity.

wangxin replied:
Agree. Using an issue to track ASIC_NOT_SUPPORTED does not make sense to me. We need to cleanup reference to such issues in conditional mark yaml file, replace with explicit condition checking asic type.

ZhaohuiS commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,517 @@
+| **Complexity** | Low | Medium | High |
+| **Offline Operation** | Yes | Yes | Limited (needs API cache) |
+| **Multi-branch Support** | Native (file per branch) | Native (file per branch) | Requires tagging/cloning |
+| **Bulk Operations** | Manual per-entry | Pattern-based | Via GitHub issue queries |
```
If someone reopens a closed issue to re-skip a test, that resets the skip with no PR review and no expiry reset. What prevents repeated reopen/close cycles from keeping a "temporary" skip alive forever? Seems like this needs either a reopen limit or a rule that reopen requires a new expiry label.

opcoder0 replied:
This is a good point and has a high possibility of occurring. I have updated the design with this scenario, and the plan is to keep such cases in the skip issue report so that it is reviewed by the leads / owners. Once an issue has reached P0 tag and has not been addressed the issue will not be removed from the report until fixed permanently.

ZhaohuiS reviewed (commented):
A few questions on the logic.

opcoder0 committed:
Fix issue expiry workflow section with correct strategy

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

mssonicbld commented:
/azp run

opcoder0 committed:
Add note to describe tool behaviour for issue re-open cases

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

mssonicbld commented:
/azp run

opcoder0 committed:
Add implementation behavior, auto-close case, cleanup sections

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

mssonicbld commented:
/azp run

opcoder0 committed:
Update design

Consider xfail as temporary, update reporting workflow, add conditional mark
edit PR workflow

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

mssonicbld commented:
/azp run

opcoder0 committed:
cleanup move older approaches to appendix

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

mssonicbld commented:
/azp run

opcoder0 committed:
cleanup add good practice

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

mssonicbld commented:
/azp run

wangxin commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,606 @@
+- Minimal disruption to existing workflows
+- Clear reporting of expired and miscategorized skips
+
+**NOTE** The conditional mark files support both `skip:` and `xfail:`. For governance and migration:
```
Indeed, conditional mark supports adding any pytest mark. Currently it is only being used as adding `skip` and `xfail` mark.

opcoder0 committed:
Modify design to do soft handling of expired issues

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

opcoder0 committed:
Modify design to allow grace period allowance to expired issues

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

mssonicbld commented:
/azp run

opcoder0 committed:
Clarify notion of temporary and permanent issue

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

opcoder0 committed:
Update design to take care of issues with no assignees by flagging in report

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

mssonicbld commented:
/azp run

wangxin commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,664 @@
+
+#### Automation Components
+
+**0. PR Review Gate Workflow (Event-Based)**
```
There is already a Github workflow to auto assign reviewers based on git history of the touched files. But the behavior of the existing workflow does not fully consistent with the description here. To implement what is described here, maybe improve the existing workflow instead of creating a new one.

opcoder0 replied:
Updated the document to indicate re-use.

wangxin commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,664 @@
+- `N` is interpreted as calendar days of temporary extension starting from the moment that label was applied (i.e., `extension_due_date = label_applied_at + N`)
+- The workflow must validate that the label was applied by a configured skip maintainer before honoring it in reporting
+- A maintainer may add the label multiple times as each extension period runs out (e.g., add `skip-maintainer-extension-30`, and after 30 days add it again to grant another 30-day window)
+- Because the same label name can appear multiple times in the issue timeline, the workflow must use the `labeled` event timestamp from the GitHub issue timeline API for each occurrence rather than relying solely on currently attached labels
```
Do you mean each "add label event" count?
What if a maintainer accidentally added a label and them removed it?
I feel that if I am a maintainer, it would be trickly for me to figure out when the issue will be actually expired simply by looking at the issue.

opcoder0 replied:
Yes, that's right. The reporting has this information. In addition

- I have also modified the implementation behavior of the workflow to add a comment on the issue every time a label is added by the workflow. The comment will indicate the next expiry date (or date of next escalation). Also added is the behavior logic to support various addition / deletion scenarios.

These are updated in the sections "Issue Comment Generation" and "Implementation Behavior"

wangxin commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,664 @@
+- The workflow must validate that the label was applied by a configured skip maintainer before honoring it in reporting
+- A maintainer may add the label multiple times as each extension period runs out (e.g., add `skip-maintainer-extension-30`, and after 30 days add it again to grant another 30-day window)
+- Because the same label name can appear multiple times in the issue timeline, the workflow must use the `labeled` event timestamp from the GitHub issue timeline API for each occurrence rather than relying solely on currently attached labels
+- An extension is considered **active** only if `today < label_applied_at + N`; expired extension occurrences are disregarded
```
For a human looking at the issue, it would be trickly to figure out when the issue will be expired.

opcoder0 replied:
Have addressed it from previous comment.

wangxin commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,664 @@
+- A designated skip maintainer may grant a grace period by adding a label in the form `skip-maintainer-extension-N`
+- `N` is interpreted as calendar days of temporary extension starting from the moment that label was applied (i.e., `extension_due_date = label_applied_at + N`)
+- The workflow must validate that the label was applied by a configured skip maintainer before honoring it in reporting
+- A maintainer may add the label multiple times as each extension period runs out (e.g., add `skip-maintainer-extension-30`, and after 30 days add it again to grant another 30-day window)
```
Bandwidth of maintainers could be a bottleneck.

wangxin commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,664 @@
+- Take action on expired issues:
+  - Add expiry-related labels (e.g., `sonic-wf-priority-3`, `sonic-wf-priority-2` etc.)
+  - Bump priority labels if using escalation strategy
+  - Add an explicit expired label when P0 duration is exceeded
```
So, the issue will be kept open? Only add a label to indicate that it is expired + mention the skip maintainers?
In reality, I am afraid that the expired issue would still be ignored.

opcoder0 replied:
This was updated based on suggestion from @StormLiangMS to not have a hard approach that could break the community's PR checkers. We only tag and the issue get reviewed in community meetings and internally.

wangxin commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,664 @@
+  Use a single issue with branch-specific labels (e.g., `branch:main`, `branch:202311`):
+  - The conditional mark plugin checks if the issue has the label for the current branch
+  - Expiry automation manages labels per branch independently
+  - Closing an issue removes it from all branches; removing a branch label removes it from that branch only
```
Assume an issue is fixed on main branch only, and 202511 branch does not have the fix.
Shall the issue be open or closed? 

And when a fix is back ported to fix the issue on a specific PR, how to ensure the corresponding label is removed from that issue?

Over time, the label definitely couldn't accurately reflect whether the issue is fixed or not on the branch.

I think this approach would be extremely difficult to maintain the accurate issue status for branches.

opcoder0 replied:
Based on our earlier discussion, we had chosen the approach to use different issues for different branches [as seen here](https://github.com/sonic-net/sonic-mgmt/pull/21495/changes/b767d758eeba3b851813b8781873f57d8288f5ae#diff-27befd8f6ef84aaa528648673f991a5a77b558f8d172320b4f26201bf68284bbR335)

wangxin commented on docs/test_skip_expiry_design.md:
```diff
@@ -0,0 +1,664 @@
+  2. The conditional mark file in that branch references the cloned issues
+  3. Each branch manages its skip lifecycle independently
+
+  This approach ensures complete isolation between branches - fixing an issue on `main` does not affect the skip status on release branches.
```
The major problem of this approach is that creating a new branch would be heavy. Need to replicate issues for the new branch, update the conditional yaml file to use the new issues.

opcoder0 replied:
Yes, I think we discussed this earlier. My initial suggestion was that since large open-source projects use tags to maintain issue fix status across branches instead of cloning issues we could do the same. After further discussion we agreed to maintain separate issues across relevant release branches. Maybe we can discuss again if required.

wangxin commented:
Overall the design looks good to me. The key points are:
1. Use workflow to update issue:
  - comment or `@` mention
  - update labels
  - notify assignee
2. Dashboard/report of issues in conditional mark

For the `update labels` part, it is a little bit too complicated to me. And updating the labels is just to affect the report/dashboard. It would be better if it could be simplified a little bit.

opcoder0 committed:
Add comment generation and tag deletion scenarios

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

mssonicbld commented:
/azp run

opcoder0 commented:
Thank you @wangxin for taking the time to review and posting the comments. I have addressed some of the comments and updated the document and commented for a few others.

StormLiangMS commented:
The design has gone through extensive review and the core direction (GitHub issue-based expiry) is solid. To unblock progress, I suggest we scope the v1 implementation to keep things pragmatic:

1. **Simple time-based expiry (Strategy A)** — Drop the P3→P2→P1→P0 escalation for now. One threshold (e.g., 6 months from issue creation), one `skip-wf-expired` label. We can add graduated escalation in v2 if needed. This addresses @wangxin's concern that the priority label system is too complex for humans to reason about.

2. **Dashboard/report first** — Build the reporting tool that scans conditional mark YAMLs and shows permanent vs temporary vs expired buckets. This alone provides huge visibility and is the most impactful deliverable.

3. **Defer branch management automation** — Handle release branch issue cloning manually for now; automate it in a follow-up PR. This avoids the heavy upfront cost @wangxin flagged.

4. **Add a migration/rollout section** — We need a concrete plan for existing permanent skips that have tracking issues today (ZhaohuiS's point). Those would be misclassified as temporary under the current rule. A one-time cleanup task or an explicit opt-out mechanism should be documented.

5. **Soft enforcement + meeting review** — The soft approach (label + @-mention, no CI blocking) is fine for v1, but let's explicitly state that all expired items will be reviewed in the weekly SONiC Test Subcommittee meeting as the escalation path.

This simplifies the design, addresses the open concerns, and lets us ship something useful quickly. @opcoder0 @wangxin thoughts?

yxieca commented:
Review note (non-blocking):
- Docs-only change; policy: do not approve.

AI agent on behalf of Ying.
<!-- AI_AGENT_REVIEW:b767d758eeba3b851813b8781873f57d8288f5ae -->

opcoder0 committed:
Update design

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

opcoder0 committed:
Restructure document

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

opcoder0 committed:
Update TOC

Signed-off-by: opcoder0 <110003254+opcoder0@users.noreply.github.com>

mssonicbld commented:
/azp run

StormLiangMS reviewed (approved):
Overall the design direction is solid and ready to move forward. Approving with a few minor items to clean up during implementation:

1. **Strategy A vs B residual inconsistency** — Line 97 picks Strategy A (time-based), but report fields at line 223 still reference Strategy B concepts (days_to_p0, current stage, priority label periods). Clean these up to use days_to_expire consistently.

2. **Appendix A is too detailed for a non-chosen strategy** — The Priority Tag Escalation appendix has 42 lines of implementation detail (comment generation, maintainer extension labels, timeline API). Since Strategy A is the decision, either trim this or clearly label it as "v2 consideration" to avoid confusion.

6. **Problem Statement still says "category-based" (line 54, 62-64)** — The final design dropped explicit categories per @wangxin's feedback (issue = temporary, no issue = permanent). Update the Problem Statement to match the chosen approach.

These are non-blocking — iterate during implementation. LGTM.
